Fix for infinite loop on Create Order page in Admin Area.
The loop is when clients add products to cart, select CashOnDelivery method and then admin disables one of the products.

diff --git a/Model/Total/Quote/AbstractTotal.php b/Model/Total/Quote/AbstractTotal.php
index 7202b2f..083751b 100644
--- a/Model/Total/Quote/AbstractTotal.php
+++ b/Model/Total/Quote/AbstractTotal.php
@@ -20,20 +20,19 @@
 
 namespace MSP\CashOnDelivery\Model\Total\Quote;
 
-use Magento\Quote\Api\PaymentMethodManagementInterface;
 use Magento\Quote\Model\Quote\Address\Total\AbstractTotal as MageAbstractTotal;
 use Magento\Quote\Model\Quote;
 
 abstract class AbstractTotal extends MageAbstractTotal
 {
     /**
-     * @var PaymentMethodManagementInterface
+     * @var \Magento\Payment\Model\MethodList
      */
-    private $paymentMethodManagement;
+    protected $methodList;
 
-    public function __construct(PaymentMethodManagementInterface $paymentMethodManagement)
+    public function __construct(\Magento\Payment\Model\MethodList $methodList)
     {
-        $this->paymentMethodManagement = $paymentMethodManagement;
+        $this->methodList = $methodList;
     }
 
     /**
@@ -47,7 +46,7 @@ abstract class AbstractTotal extends MageAbstractTotal
         if (!$quote->getId()) {
             return false;
         }
-        $paymentMethodsList = $this->paymentMethodManagement->getList($quote->getId());
+        $paymentMethodsList = $this->methodList->getAvailableMethods($quote);
         if ((count($paymentMethodsList) == 1) && (current($paymentMethodsList)->getCode() == 'msp_cashondelivery')) {
             return true;
         }
diff --git a/Model/Total/Quote/Cashondelivery.php b/Model/Total/Quote/Cashondelivery.php
index 224aa01..15a89c5 100644
--- a/Model/Total/Quote/Cashondelivery.php
+++ b/Model/Total/Quote/Cashondelivery.php
@@ -34,11 +34,11 @@ class Cashondelivery extends AbstractTotal
     protected $priceCurrencyInterface;
 
     public function __construct(
-        PaymentMethodManagementInterface $paymentMethodManagement,
+        \Magento\Payment\Model\MethodList $methodList,
         PriceCurrencyInterface $priceCurrencyInterface,
         CashondeliveryInterface $cashOnDeliveryInterface
     ) {
-        parent::__construct($paymentMethodManagement);
+        parent::__construct($methodList);
 
         $this->cashOnDeliveryInterface = $cashOnDeliveryInterface;
         $this->priceCurrencyInterface = $priceCurrencyInterface;
diff --git a/Model/Total/Quote/CashondeliveryTax.php b/Model/Total/Quote/CashondeliveryTax.php
index b9d23e3..23da079 100644
--- a/Model/Total/Quote/CashondeliveryTax.php
+++ b/Model/Total/Quote/CashondeliveryTax.php
@@ -34,11 +34,11 @@ class CashondeliveryTax extends AbstractTotal
     protected $priceCurrencyInterface;
 
     public function __construct(
-        PaymentMethodManagementInterface $paymentMethodManagement,
+        \Magento\Payment\Model\MethodList $methodList,
         PriceCurrencyInterface $priceCurrencyInterface,
         CashondeliveryInterface $cashOnDeliveryInterface
     ) {
-        parent::__construct($paymentMethodManagement);
+        parent::__construct($methodList);
         $this->cashOnDeliveryInterface = $cashOnDeliveryInterface;
         $this->priceCurrencyInterface = $priceCurrencyInterface;
     }
